### PR TITLE
Paper: Performing Object Detection on Drone Orthomosaics with Meta's Segment Anything Model (SAM)

### PR DESCRIPTION
**Pull request by nickmccarty**
Pushed changes to pass doi checks, please advise if anything else is required...

### CODE REVIEW

**Comment by ameyxd** [comment]
Inviting reviewers: @mainaksingha.iitb@gmail.com and @naomitangyuan@gmail.com

**Comment by nickmccarty** [comment]
Is the accompanying poster to be submitted using MyST/Curvenote as well, @scipy-conference/2025-proceedings? If so, are you able to point me toward any guidance on that front?

**Comment by nickmccarty** [comment]
Are you able to field this inquiry, @ameyxd? 👆

**Comment by ameyxd** [comment]
> Are you able to field this inquiry, @ameyxd? 👆

paging @rowanc1 / @fwkoch

**Comment by rowanc1** [comment]
Please open another PR with your poster and follow the [instructions in the readme](https://github.com/scipy-conference/scipy_proceedings?tab=readme-ov-file#instructions-for-slides)!

**Comment by nickmccarty** [comment]
> Please open another PR with your poster and follow the [instructions in the readme](https://github.com/scipy-conference/scipy_proceedings?tab=readme-ov-file#instructions-for-slides)!

Will do, thanks a bunch!

**Comment by ameyxd** [comment]
@anacomesana will serve as editor for this paper.

**Comment by naomity** [comment]
Hi Nick, it's my honor to be assigned to review this paper. The paper is overall well-structured, will be taking a closer look soon!

File: papers/nicholas_mccarty/myst.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+version: 1
+extends: ../papers.yml
+project:
+  # Update this to match `scipy-2025-<folder>` the folder should be `<firstname_surname>`
+  id: scipy-2025-nicholas_mccarty
+  # Ensure your title is the same as in your `main.md`
+  title: Performing Object Detection on Drone Orthomosaics with Meta's Segment Anything Model (SAM)
+  # subtitle:
+  description: This article presents a workflow that utilizes SAM's automatic mask generation skill to effectively perform the task of object detection zero-shot on a high-resolution drone orthomosaic. The generated output is 20% more spatially accurate than that produced using proprietary software, with 400% greater IoU.
```
Comment (naomity) [inline papers/nicholas_mccarty/myst.yml:9]: `20% more spatially accurate, with 400% greater IoU`. I assume this is comparing with vanilla SAM?

File: papers/nicholas_mccarty/main.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
... presents a fully open source workflow leveraging Meta AI's Segment Anything Model (SAM) for zero-shot segmentation, enabling scalable object detection and spatial localization in high-resolution drone orthomosaics without the need for annotated image datasets. Model training and/or fine-tuning is rendered unnecessary in our precision agriculture-focused use case. The presented end-to-end workflow takes high-resolution images and quality control (QC) check points as inputs, automatically generates masks corresponding to the objects of interest (empty plant pots, in our given context), and outputs their spatial locations in real-world coordinates. Detection accuracy (required in the given context to be within 3 cm) is then quantitatively evaluated using the ground truth QC check points and benchmarked against object detection output generated using commercially available software. Results demonstrate that the open source workflow achieves superior spatial accuracy — producing output `20% more spatially accurate`, with `400% greater IoU` — while providing a scalable way to perform spatial localization on high-resolution aerial imagery (with ground sampling distance, or GSD, < 30 cm).
```
Comment (naomity) [inline papers/nicholas_mccarty/main.md:5]: ditto, `20% more spatially accurate, with 400% greater IoU`. I assume this is comparing with vanilla SAM?

File: papers/nicholas_mccarty/main.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+:::{figure} trimmer.jpg
+:label: fig:trimmer
+The derived centroids of the objects detected in the drone orthomosaic are used to automate this nursery trimmer.
+:::
+
+SAM’s zero-shot segmentation capability directly addresses the data annotation bottleneck, enabling rapid deployment in novel contexts. By developing an open source workflow around SAM, an end-to-end pipeline is created which allows for the quantitative evaluation of spatial accuracy with respect to objects detected in high-resolution aerial imagery. This modular workflow can also be repurposed as an automated data annotation pipeline for downstream model training/fine-tuning, if required.
+
+## Approach
+
+Our approach integrates SAM’s segmentation strengths with traditional geospatial data processing techniques, which lends itself to our precision agriculture use case. The workflow, like any other, can be thought of as a sequence of steps (visualized above and described below), each with their own sets of substeps:
```
Comment (naomity) [inline papers/nicholas_mccarty/main.md:42]: nit: `visualized above`, missing / mis-placed visualization?

File: papers/nicholas_mccarty/mask-generation.png
Comment (naomity) [inline papers/nicholas_mccarty/mask-generation.png:1]: Curious why is 80% used as threshold? Is it a hyper-paramter for optimized performance, or a standard practice?

File: papers/nicholas_mccarty/main.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+## Discussion
+
+### Key Findings
+
+The open-source workflow using Meta AI’s Segment Anything Model (SAM) outperformed a commercial alternative in object detection and spatial localization on high-resolution drone imagery. It achieved `20% higher spatial accuracy` (1.20 cm vs 1.39 cm deviation) and a `400% higher Intersection-over-Union (IoU)` (0.74 vs 0.18), indicating stronger alignment with object boundaries. Both methods had near-perfect precision, but the open-source approach showed slightly lower recall due to 65 false negatives. It should be noted, however, that these FN were a direct result of the filtering substep in our workflow, which filtered our detections (based on arbitrary geometry area and compactness thresholds; see [code](https://colab.research.google.com/drive/1pwnb14s2i7n_VAlfwhBqzDQ0cOb9oGs-?usp=sharing#sandboxMode=true&scrollTo=240nXaT5-EqM)) that are present in the output.
```
Comment (naomity) [inline papers/nicholas_mccarty/main.md]: Can author elaborate that, from a domain expert point of view, which metric(s) is the most important and why?

File: papers/nicholas_mccarty/main.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+  - Rasterio [@rasterio]
+  - Segment Anything[^footnote-3] [@kirillov23]
+  - Shapely [@shapely]
+  - Torch [@pytorch]
+
+[^footnote-2]: See [requirements.txt](https://raw.githubusercontent.com/nickmccarty/scipy-2025/refs/heads/main/requirements.txt) for version details.
+
+[^footnote-3]: Inference was accelerated using `CUDA 12` (`cuDF 25.2.1`) on a `T4` GPU within our Colab notebook environment.
+
+### Workflow
```
Comment (naomity) [inline papers/nicholas_mccarty/main.md:83]: For the Workflow section, can we replace or add a pseudocode-algorithm style block? It will be a more formal and technical representation.

**Review by naomity — COMMENTED** [review]
Impressive dataset and application domain! Paper is well-written, just a few minor edits.

File: papers/nicholas_mccarty/myst.yml (hunk 1 above)
Comment (nickmccarty) [reply papers/nicholas_mccarty/myst.yml:9]: Good question, which indicates that I need to add clarity on this front. We benchmarked our results against the output produced using proprietary software; our output is more spatially accurate (the centers of our detected objects are closer to the QC points) and our polygons cover the actual objects better (our generated mask polygons have 400% greater IoU than the bounding boxes generated using the proprietary software). Will add this to my list of edits.

File: papers/nicholas_mccarty/main.md (hunk 2 above)
Comment (nickmccarty) [reply papers/nicholas_mccarty/main.md:5]: See above comment for clarification.

File: papers/nicholas_mccarty/main.md (hunk 3 above)
Comment (nickmccarty) [reply papers/nicholas_mccarty/main.md:42]: Good catch, will edit!

File: papers/nicholas_mccarty/mask-generation.png
Comment (nickmccarty) [reply papers/nicholas_mccarty/mask-generation.png:1]: Empirically, through iteration, we discovered that threshold produced more useful results.

File: papers/nicholas_mccarty/main.md (hunk 4 above)
Comment (nickmccarty) [reply papers/nicholas_mccarty/main.md]: Spatial accuracy was a metric imposed by the client, and IoU is a standard evaluation metric in computer vision/object detection.

File: papers/nicholas_mccarty/main.md (hunk 5 above)
Comment (nickmccarty) [reply papers/nicholas_mccarty/main.md:83]: Good callout, will do!

File: papers/nicholas_mccarty/main.md [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+## AI Usage Disclosure
+
+AI tools (ChatGPT, Perplexity, and NotebookLM) were used:
+
+- in writing portions of the workflow integration code,
+- to generate Matplotlib subplots, process flow diagrams, <span style="font-family: serif;">L<span style="vertical-align: 0.4ex; font-size: 0.8em;">A</span>T<span style="vertical-align: -0.3ex; font-size: 0.8em;">E</span>X</span>, etc.
+- for proofreading and light revision to reduce potential publication errors.
+
+## Code
```
Comment (fwkoch) [inline papers/nicholas_mccarty/main.md:289]:
```suggestion
(code)=
## Code
```

File: papers/nicholas_mccarty/main.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+Data and code required to replicate our approach can be found using the links below:
+
+[![GitHub](https://img.shields.io/badge/github-%23121011.svg?style=for-the-badge&logo=github&logoColor=white)](https://github.com/nickmccarty/scipy-2025)
+<a href="https://colab.research.google.com/drive/1pwnb14s2i7n_VAlfwhBqzDQ0cOb9oGs-?usp=sharing#sandboxMode=true">
+  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
+</a>
+
+## Appendix
+
+### Accuracy Evaluation Methodology
```
Comment (fwkoch) [inline papers/nicholas_mccarty/main.md:301]:
```suggestion
(accuracy-evaluation-methodology)=
### Accuracy Evaluation Methodology
```

**Review by fwkoch — COMMENTED** [review]
Thanks for coming by our booth 🙂 Let's see if these changes get the PDF building...

**Comment by nickmccarty** [comment]
> Thanks for coming by our booth 🙂 Let's see if these changes get the PDF building...

Thanks a bunch -- you rock! Great chatting with you, Franklin 🤘🤩

**Comment by fwkoch** [comment]
I noticed there are a few sections where you only have a figure, and in the PDF, these figures get displaced and the sections look empty (see the image below).

We can also get this fixed up for you - there is an option to anchor figures in place. I need to determine if we should make this fix across all papers or just yours. Don't worry about it for now!

image

**Comment by nickmccarty** [comment]
> I noticed there are a few sections where you only have a figure, and in the PDF, these figures get displaced and the sections look empty (see the image below).
> 
> We can also get this fixed up for you - there is an option to anchor figures in place. I need to determine if we should make this fix across all papers or just yours. Don't worry about it for now!
> 
> image

Thank you for the consideration! The reviewer mentioned that replacing the figures with pseudocode would be better, so I'll likely do that. Will circle back if I decide otherwise -- thanks again, @fwkoch!

**Comment by nickmccarty** [comment]
I pushed the suggested revisions and added labels to the various sections (when linking) like you did, but the changes are somehow preventing the PDF from building again, @fwkoch ... a million thanks in advance for any help or guidance you could provide!

File: papers/nicholas_mccarty/main.md [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+:label: fig:qc-point-91
+Visual inspection of the detected centroids relative to QC point 91 reveal that the QC point is off-center.
+:::
+
+Visual inspection also reveals that our detections (the pink circle) and those produced using the commercial software (the beige square) have greater overall coverage with respect to the QC geometry (the grey square). This provides intuition as to why the IoU calculations revealed a 400% increase in coverage with respect to the geometry produced using SAM's automatic mask generator, zero-shot.
+
+## Discussion
+
+### Key Findings
+
```
Comment (anacomesana) [inline papers/nicholas_mccarty/main.md:245]: It would be helpful to clarify that the large IoU improvement arises not only from quantitative differences but also from the methodological distinction between mask-based polygons and bounding boxes. A brief reminder of the different output types would make this clearer. 
In addition, since the FN stem from the filtering substep, consider noting whether you observed practical tradeoffs between FP and FN, and that these thresholds were empirically chosen. A short clarification would strengthen the interpretation of the results.

Comment (nickmccarty) [reply papers/nicholas_mccarty/main.md:245]: Thanks for your input, @anacomesana  -- I'll be sure to try to get those points added ahead of tomorrow's deadline (I'm traveling) 🤞 Also, while I have you here, do I need to alter the Methodolody pseudocode by instead using the LaTeX `algpsedocode` package, or anything? Thanks again!

File: papers/nicholas_mccarty/main.md [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+Nevertheless, its overall performance supports its suitability for precision agriculture and downstream automation.
+
+### Precision Agriculture Challenges
+
+Our work began with an eye toward tackling a major challenge in agricultural remote sensing: the need for extensive manual annotation. SAM’s zero-shot segmentation enables accurate object detection without domain-specific training, making it scalable and adaptable for new use cases with minimal setup.
```
Comment (anacomesana) [inline papers/nicholas_mccarty/main.md:252]: Consider adding a short reflection here on the workflow's potential generalizability (for example, if similar performance could be expected in other agricultural imagery like crop rows or tree canopies), or if additional challenges might arise in different applications.

Comment (nickmccarty) [reply papers/nicholas_mccarty/main.md:252]: Again, thanks -- I'll try ahead of tomorrow's deadline (not a lot of advance notice)

Also, these are trade secrets -- this is not an academic project. I have a responsibility to my client to say as little as possible (I've written and submitted this with their permission and their extended grace)...

**Comment by nickmccarty** [comment]
Will I be penalized if I don't make the reviewer's last-minute suggested revisions, @scipy-conference/2025-proceedings? If so, I'm traveling and would like to request some grace, given the lack of notice w.r.t. these oddly-timed (questionable) revision requests...

**Comment by nickmccarty** [comment]
Also, can I please get the requested guidance about the pseudocode, @anacomesana?

**Review by naomity — APPROVED** [review]

**Comment by ameyxd** [comment]
@nickmccarty and @anacomesana - please resolve this today (09/10/2025) to ensure timely publication.

**Comment by nickmccarty** [comment]
> @nickmccarty and @anacomesana - please resolve this today (09/10/2025) to ensure timely publication.

Would love a response from my reviewer...

**Comment by nickmccarty** [comment]
👆 @scipy-conference/2025-proceedings

**Comment by nickmccarty** [comment]
If I don't get relevant, actionable input today, will my paper be published, or no? I would like a response please, as I am mid-flight to Hawaii to celebrate my wife's birthday and have planned my day accordingly...

**Comment by hongsupshin** [comment]
Hi @nickmccarty, sincere apologies from the Proceedings committee for the delayed response. Your reviewer and my fellow co-chair @anacomesana had to handle personal matters that have caused the delay. @anacomesana will finish the review in time, and we will make sure your paper gets published. Please do not worry about the deadline. Thanks for your patience and enjoy your travel!

File: papers/nicholas_mccarty/main.md [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+
+### Practical Impact
+
+Meeting professional-grade tolerance requirements (e.g., < 3 cm) enables real-world applications, such as automating heavy equipment, based on precise object localization. This demonstrates how automated workflows can reduce manual labor and support more efficient agricultural practices.
+
+### Limitations and Future Work
+
+Our approach to tiling ("chipping") high-resolution orthomosaics, processing 588 individual 1280-by-1280px tiles at an average pace of 11 seconds per tile, required a total processing time of ~110 minutes running on a Colab single `T4` GPU instance. It is important to note that an overlap of 25% (320px) between tiles during processing was required to ensure that geometry was not produced containing "holes" or malformations; merging overlapping polygons after filtering (based on area and compactness calculations, in this case) helped us ensure the overall quality of the geometric output.
+
+Future work will be centered on building an open source CLI and Python package[^footnote-6], which will allow users to pass orthomosaics as inputs and get geometry meeting desired spatail charactersitics as an output.
```
Comment (anacomesana) [inline papers/nicholas_mccarty/main.md]: minor typo! "spatail" instead of "spatial"

File: papers/nicholas_mccarty/main.md [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+[^footnote-5]: We used open source QGIS [@qgis] as our selected data viewer.
+
+### Open Source Workflow
+
+:::{figure} open-source-false-positive-collage.png
+:label: fig:open-source-workflow-output-evaluation
+18 FP were observed in the output produced using the open source workflow. 
+:::
+
+Merging the overlapping geometry and filtering out the empircially observed FP allowed for us to ascertain exactly how many TP (18,736) there are in the benchmark output and derive how many FP (18) and FN (65) there are in our workflow output, which enabled us to conduct our performance comparison.
```
Comment (anacomesana) [inline papers/nicholas_mccarty/main.md]: minor typo! "empircially"" instead of "empirically"

**Comment by anacomesana** [comment]
Hi @nickmccarty, apologies for my delay here, and thank you for your patience. I've gone through the paper carefully. 
In terms of the methodology, the current numbered workflow is clear and works as-is.
I caught two minor typos, and once those are fixed, I believe the paper will be in great shape for publication!

Thank you so much, and my apologies once again for the delay.
